### ISSUE TITLE
Different behaviour for rtcp-mux: require from 7.2.1.2+0~mr7.2.1.2 to current master 7.5.2.6+0~mr7.5.2.6

### ISSUE THREAD

**Issue by elbow** [sipwise/rtpengine#957]
Both compiled from source by building the debs

I have the following SDP - which is a response from upstream SIP server (Asterisk 13) after receving an "a=sendonly" to place party on hold.

```
v=0
o=root 1860435649 1860435650 IN IP4 a.b.c.133
s=Telviva
c=IN IP4 a.b.c.133
t=0 0
m=audio 18912 RTP/AVP 9 8 126
a=rtpmap:9 G722/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:126 telephone-event/8000
a=fmtp:126 0-16
a=ptime:20
a=maxptime:150
a=recvonly
```

I do thusly:

```
2020-03-26T13:41:54.739Z rtpengine:client RtpEngine: sending command: 5f9ba5c8-7e0f-4a46-963c-1d78d6e4e07a: {"call-id":"uc4si49jhqgh2rdkcnje","from-tag":"dt38u5ldo1","ICE":"force","SDES":"off","DTLS":"passive","flags":["SD
ES-no"],"codec":{"strip":["telephone-event"],"mask":["all"],"transcode":["opus"],"offer":["telephone-event"]},"rtcp-mux":["require"],"transport-protocol":"UDP/TLS/RTP/SAVPF","sdp":"v=0\r\no=root 1860435649 1860435650 IN IP
4 a.b.c.133\r\ns=Telviva\r\nc=IN IP4 a.b.c.133\r\nt=0 0\r\nm=audio 18912 RTP/AVP 9 8 126\r\na=rtpmap:9 G722/8000\r\na=rtpmap:8 PCMA/8000\r\na=rtpmap:126 telephone-event/8000\r\na=fmtp:126 0-16\r\na=ptime:20\r\n
a=maxptime:150\r\na=recvonly\r\n","to-tag":"33B4mD44ap7Zp","command":"answer"}`
```

RTPengine gives me back:

```
v=0
o=root 1860435649 1860435650 IN IP4 a.b.c.133
s=Telviva
c=IN IP4 a.b.c.157 (that's the rtpengine boxes ip)
t=0 0
m=audio 42520 UDP/TLS/RTP/SAVPF 9 8 126
a=maxptime:150
a=mid:0
a=rtpmap:9 G722/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:126 telephone-event/8000
a=fmtp:126 0-16
a=recvonly
a=rtcp:42521
a=setup:passive
a=fingerprint:sha-1 6F:90:15:06:D3:1E:96:00:47:E5:F8:89:EC:17:81:6D:EA:7C:4A:D3
a=ptime:20
a=ice-ufrag:L8dfRI2W
a=ice-pwd:ldpAQFQy31Dl9nrbTPWCOCnf35
a=ice-options:trickle
a=candidate:Z0SWFZwZwz2mwaPb 1 UDP 2130706431 a.b.c.157 42520 typ host
a=candidate:Z0SWFZwZwz2mwaPb 2 UDP 2130706430 a.b.c.157 42521 typ host
a=end-of-candidates
```

As you can see with the latest version rtpengine does not setup the rtcp-mux.

This results in Chrome rejecting this SDP and the loss of the call.

The older version does do the rtcp-mux and hold works as expected.

Thanks for your assistance, let me know if I need to provide more info.

**Comment by rfuchs** [comment]
Can you post the full call log including previous offers/answes?

**Comment by elbow** [comment]
[hold-issue.txt](https://github.com/sipwise/rtpengine/files/4388162/hold-issue.txt)

Hi!

I've extracted a full trace start to finish including the various calls to rtpengine.

Thanks

**Comment by rfuchs** [comment]
Is there a B2BUA in between the two endpoints? Or something else that substitutes the to-tag? I see both `as6db24c36` and `33B4mD44ap7Zp` used as to-tag.

**Comment by elbow** [comment]
Yes, at the SIP level the arthitecture is back-to-back.

The sip engine is Drachtio a SIP platform using Sofia which enables coding application logic with Javascript and node. 

Obviously Rtpengine doesn't see the SIP stuff. 

Rtpengine is in the middle for audio.

**Comment by rfuchs** [comment]
Rtpengine does see the tags though, and it sees two different tags used for SIP/RTP side of things. You can't do that - the tags have to be consistent. One tag = one side of a dialogue.

FTR: rtcp-mux=require doesn't actually do anything in an answer. It's only useful in an offer, and only in an initial offer for that matter, because rtpengine remembers whether rtcp-mux is supported by an endpoint or not. So: initial offer from WebRTC to SIP: rtcp-mux=demux. Or initial offer the other direction: rtcp-mux=require. That's all that's required.

**Comment by elbow** [comment]
Hi Richard,

Thanks for the help. Were we just lucky with the older version in some way, or was this tightened up or changed? 

Since the tags are not in the actual sdp, I presume this means I can just pick a from tag - like from the original A side invite - and a to-tag - like the one we get on the B side, and use them consistently? 

Or is it legit to reuse the from tag of the A side as the from tag on the Invite on the B side? And similarly on the way back. Can I get away with that since the call Id is different so the dialogs should stay distinct? 

I'm feeling unsure due to the complexities of sip and I'm not an experienced rtpengine user. 

Thank you.

**Comment by rfuchs** [comment]
If you invoke rtpengine always from just one side of the B2BUA (e.g. always on the WebRTC side), then you should automatically get the right behaviour. The tags normally taken from the SIP message being processed.

The call IDs are actually already consistent from rtpengine's POV: I only see one (`uc4si49jhqgh2rdkcnje`) in the log. And this is how it needs to be. But for some reason, one of the to-tags from the other side of the B2B (`as6db24c36` from call ID `IyODBAc2l~rp~uc4si49jhqgh2rdkcnje`) is also given to rtpengine in some of the answers, albeit with the other call ID `uc4si49jhqgh2rdkcnje`. It should have been given the corresponding tag from the `uc4si49jhqgh2rdkcnje` leg of the call, which is `33B4mD44ap7Zp` instead. I'm not sure how a call ID from one SIP message gets mixed up with a tag from a different SIP message, unless you supply them manually.

So from rtpengine's POV you need: 1) always the same call ID, 2) always the same tag for the A party, 3) always the same tag for the B party.

**Comment by elbow** [comment]
Thanks for taking the time to help me understand.  It is greatly appreciated.

Original invite: `"call-id":"uc4si49jhqgh2rdkcnje","from-tag":"dt38u5ldo1","command":"offer"`

Authenticated invite: `"call-id":"uc4si49jhqgh2rdkcnje","from-tag":"dt38u5ldo1","command":"offer"`

The 200 OK: `"call-id":"uc4si49jhqgh2rdkcnje","from-tag":"dt38u5ldo1",to-tag":"as6db24c36","command":"answer"`

So we used the callid from the rtc leg, the from tag from the rtc leg and the to tag from the upstream leg.

The sendonly re-INVITE: `"call-id":"uc4si49jhqgh2rdkcnje","from-tag":"dt38u5ldo1","command":"offer"`

So that is consistent

The 200 OK for the re-INVITE: `"call-id":"uc4si49jhqgh2rdkcnje","from-tag":"dt38u5ldo1","to-tag":"33B4mD44ap7Zp","command":"answer"`

Here we dropped the ball - since we now (somehow!) used the to tag from the rtc leg rather than the to tag from the upstream sip leg.
 
So yes, it does look like we went to some trouble to do the wrong thing.

**Comment by elbow** [comment]
Yeah, found the spot where we go wrong.  I'll fix tomorrow and confirm and close the ticket.

**Comment by elbow** [comment]
Hi,

I wrangled my tags into order and I can now place calls on hold.  Thanks for the help.

I have a followup issue where I have no audio path when I hold and unhold from the JSSIP side - but only in the case where I'm holding and unholding a call that was originally inbound to JSSIP.  If I hold/unhold an outbound call I'm fine.

I'l open a separate issue for that if I can't find the problem.  I did try port-latching.
